Azure storage driver

Andrey-mp opened the pull request:
This patchset adds support of Azure Cloud.
There are several operations implemented in patchset:
create, attach, detach, remove (all base operations
for volumes).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/codedellemc/libstorage?pullRequest=372) <br/>All committers have signed the CLA.

akutz commented:
Hi @Andrey-mp,

Do you mind rebasing this brand onto `master`? Thank you!

akutz commented on Makefile:
```diff
@@ -847,6 +847,7 @@ $1:
 	BUILD_TAGS="$$(BUILD_TAGS)" GOOS=$2 GOARCH=amd64 $$(MAKE) $$@
 $1-clean:
 	rm -f $1
+	rm -f $(EXECUTORS_GENERATED)
```
Good catch!

akutz commented on drivers/storage/azure/azure.go:
```diff
@@ -0,0 +1,99 @@
+
+const (
+	// ConfigAZURE is a config key
+	ConfigAZURE                    = Name
```
Hi @Andrey-mp,

This is the only place I will mention this in my review, but why did you make `AZURE` capitalized throughout your use of it in constant/variable names? I checked, and it seems Microsoft doesn't use all-caps to write _Azure_, so I'm just wondering why you did?

![image](https://cloud.githubusercontent.com/assets/101085/22075647/7db134a0-dd72-11e6-9379-4a6f3298c24d.png)

Andrey-mp replied:
Will be renamed.

akutz commented on drivers/storage/azure/executor/azure_executor.go:
```diff
@@ -0,0 +1,141 @@
+	ctx types.Context,
+	opts types.Store) (string, error) {
+	// All possible device paths on Linux instances are /dev/sd[c-p]
+	letters := []string{
```
Since Go will keep malloc'd memory on the stack if it does not leak it probably makes more sense to keep this variable local to this function, but since mallocs are expensive in Go, I'm wondering if it makes sense to turn this into a package-scoped field that is allocated once on the heap.

akutz commented on drivers/storage/azure/executor/azure_executor.go:
```diff
@@ -0,0 +1,141 @@
+	localDeviceMapping := localDevices.DeviceMap
+
+	for localDevice := range localDeviceMapping {
+		re, _ := regexp.Compile(`^/dev/` +
```
RegEx compilations are pretty expensive. The pieces of this regex are static, so why not use `regexp.MustCompile` once and assign the result to a package-level field. Compiling the regex in a loop seems pretty expensive.

akutz commented on drivers/storage/azure/storage/azure_storage.go:
Hi @Andrey-mp,

I wasn't sure if you intended to do a case-sensitive comparison or not. If not then perhaps see [`strings.EqualFold(string, string) bool`](https://golang.org/pkg/strings/#example_EqualFold).

akutz commented on drivers/storage/azure/storage/azure_storage.go:
Hi @Andrey-mp,

I'm not going to make this comment on each of the below functions, but this remark does apply to all of the code below related to retrieving configuration properties.

There is no need to check the environment variables directly. Please see this [example](https://github.com/codedellemc/libstorage/blob/master/drivers/storage/ebs/storage/ebs_storage.go#L1321-L1332) for the EBS driver (ignore the backwards-compatability checks). The [gofig](https://github.com/akutz/gofig) package uses @spf13's [Viper](https://github.com/spf13/viper), which in turn enables the assignment of environment variables to defined configuration properties. 

Let's look at [line \#87](https://github.com/cloudscaling/libstorage/blob/f4bf3809fe1e354dbccd7ef05086f06677c7ad6b/drivers/storage/azure/azure.go#L87) in your `azure.go`:

```go
r.Key(gofig.String, "", "", "", ConfigAZURESubscriptionIDKey)
```

In addition to notifying Gofig about the property `azure.subscriptionID`, the above function call instructs Viper to bind the property `azure.subscriptionID` to the CLI flag `--azureSubscriptionID` and the environment variable `AZURE_SUBSCRIPTIONID`. You can actually use Gofig to overwrite which CLI flag or environment variable the property is bound via the [`Key` function's trailing variadic parameter](https://github.com/akutz/gofig/blob/master/gofig_reg.go#L56-L114). 

For example, if we wanted to still use the config property `azure.subscriptionID` but change the CLI flag to `azureSubID` and the environment variable to `AZURE_SUB_ID` we could rewrite the above line like so:

```go
r.Key(gofig.String, "", "", "", 
	ConfigAZURESubscriptionIDKey,
	"azureSubID",
	"AZURE_SUB_ID")
```

Once the configuration properties are bound to a CLI flag or environment variable, invoking `config.GetString(azure.ConfigAZURESubscriptionIDKey)` checks first the environment variable, CLI flag argument, and finally a configuration property in that order.

Andrey-mp replied:
Done.

akutz commented on drivers/storage/azure/storage/azure_storage.go:
Hi @Andrey-mp,

Do you intend to have this error logged but not returned? If so perhaps log it as a warning instead?

akutz reviewed (changes requested):
Hi @Andrey-mp,

This is a great PR -- thank you so very much the obvious time and effort you spent on this. I have some suggestions/questions to which I hope you're able to respond below. 

One question I had regarding the entire PR is whether or not you used the GoMetalinter or at least go-fmt to ensure the files are consistent in their syntax/style? 

Also, if possible can you try and limit the width of lines to 80 chars. I may be pedantic, but I find it makes things easier to read.

Thank you again!

codenrhoden commented on drivers/storage/azure/azure.go:
```diff
@@ -0,0 +1,99 @@
+	// ContainerKey is a name of container in the storage account ('vhds' by default)
+	ContainerKey      = "container"
+	// UseHTTPS is a flag about use https or not for making Azure URI's
+	UseHTTPS          = "useHTTPS"
```
Looks like both `UseHTTPS` and `StorageAccount` are used only when setting up config keys. To be consistent with all the other keys used in this manner, I feel like these should be renamed to `UseHTTPSKey` and `StorageAccountKey`.

codenrhoden commented on drivers/storage/azure/executor/azure_executor.go:
```diff
@@ -0,0 +1,141 @@
+
+const procPartitions = "/proc/partitions"
+
+var devRX = regexp.MustCompile(`^sd[a-z]$`)
```
This is looking for `/dev/sd[a-z]`, but the comment in `NextDevice` says it can only be `/dev/sd[c-p]`. Are these both true?

If newly attached devices can only go to `p`, do we need to include all the way up to `z`? 

Functionally I'm sure this works, it just jumps out to me as an inconsistency.

Andrey-mp replied:
Done.

codenrhoden replied:
This looks the same to me. `availLetters` is still defining [c-z], the regex is looking for [a-z], and the first comment inside of `NextDevice` says that it can be [c-p].

codenrhoden commented on drivers/storage/azure/storage/azure_storage.go:
resourceGroup doesn't seem to be a part of uniquely identifying a session. I don't think it needs to be a part of the hash key. It's never used in this function again.

codenrhoden commented on drivers/storage/azure/storage/azure_storage.go:
Seems like this can be removed for now?

codenrhoden commented on drivers/storage/azure/storage/azure_storage.go:
should an error here be ignored? What if the API failed and the check wasn't able to be performed?

codenrhoden commented on drivers/storage/azure/storage/azure_storage.go:
Is it necessary for the Azure driver to do this filtering itself? I don't think that it is, and I don't think that it needs to handle setting `attachState` either. The framework handles this for you, here: https://github.com/codedellemc/libstorage/blob/release/0.4.0/api/server/router/volume/volume_routes.go#L131-L209

If a driver really does need to calculate `attachState` itself, it can do so. That will override the framework's default behavior of saying the volume is available if there are no attachments returned, it is attached if it is attached to the IID of the caller, or unavailable if attached to a different IID.

It is the intended behavior that `attachState` is not set if `attachments.Requested()` is false.

codenrhoden commented on drivers/storage/azure/utils/utils.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+// IsAzureInstance returns a flag indicating whether the executing host is an Azure
+// instance based on whether or not the metadata URL can be accessed.
```
The comment here does not match how the function is actually implemented. It does not use metadata that I can see.

codenrhoden commented on drivers/storage/azure/utils/utils.go:
```diff
@@ -0,0 +1,77 @@
+		// UUID can be obtained as descried in https://azure.microsoft.com/en-us/blog/accessing-and-using-azure-vm-unique-id/
+		// but this code will use hostname as ID
+
+		hostname, err = os.Hostname()
```
How likely is this to happen? I'm guess that `AZURE_INSTANCE_ID` should normally be set. But falling back to a hostname seems iffy, unless Asure has requirements on whether hostnames have to be unique?

May be a non-issue, just curious.

Andrey-mp replied:
Azure doesn't set AZURE_INSTANCE_ID inside VM-s. This variable was implemented for unit tests only.
Azure set hostname inside of VM equals to name of VM in Azure engine. And this name is unique in the engine.
So because there is no other way to obtain instance identifier (that is suitable for REST API) then we get it from the hostname.

codenrhoden reviewed (changes requested):
Echo'ing what @akutz said, thanks for the very thorough and complete PR! Hurray for tests.

Most of my comments are minor. My main feedback is that I don't think the filtering in toTypeVolume is necessary, and i also think that handling of the attachmentState can be handled by the framework instead of the driver.

Andrey-mp committed:
Support Azure for libstorage

This patchset adds support of Azure Cloud.
There are several operations implemented in patchset:
create, attach, detach, remove (all base operations
for volumes).

Andrey-mp commented:
Hi all.

We've thought that we fixed all your comments :)
Also I run gometalinter and it doesn't show me valuable errors.
We are waiting for next review.

codenrhoden reviewed (changes requested):
Just have the one remaining comment on device names. I'll start playing with this on Azure!

I'd like to see a first attempt made at adding the driver into the docs, at .docs/user-guide/storage-providers.md.

Should be interesting, as I haven't used Azure before.

Andrey-mp committed:
fix comment in NextDevice

Andrey-mp commented:
Hi Travis,
We've got this code from the EBS driver that is very similar to Azure.
And in it code I see the same - NextDevice and LocalDevices has different letters...
And other executors have different behaviour for these functions also.

codenrhoden commented:
Closing in deference to #421
